### PR TITLE
[FVM] beyond EVM part 2.1 - adding EVM emulator database

### PR DESCRIPTION
**Pull request by ramtinms**
The first part of [this large PR](https://github.com/onflow/flow-go/pull/4860)

### CODE REVIEW

File: fvm/evm/emulator/database/database.go [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+}
+
+// Write flushes any accumulated data to the memory database.
+func (b *batch) Write() error {
+	return b.db.ApplyBatch(b)
+}
+
+// Reset resets the batch for reuse.
+func (b *batch) Reset() {
+	b.writes = b.writes[:0]
```
Comment (fxamacker) [inline fvm/evm/emulator/database/database.go:353]: nit: maybe reset `writes` elements to release memory if value is large.

File: fvm/evm/emulator/database/storable.go [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+		return nil, err
+	}
+
+	switch t {
+	case cbor.ByteStringType:
+		s, err := dec.DecodeBytes()
+		if err != nil {
+			return nil, err
+		}
+		return NewByteStringValue([]byte(s)), nil
```
Comment (fxamacker) [inline fvm/evm/emulator/database/storable.go:136]:
```suggestion
		return NewByteStringValue(s), nil
```

**Review by fxamacker — COMMENTED** [review]
Nice work!  I left some comments and suggestions related to atree usage.

File: fvm/evm/emulator/database/database.go [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+		m, err = atree.NewMap(db.storage, atree.Address(db.flowEVMRootAddress), atree.NewDefaultDigesterBuilder(), emptyTypeInfo{})
+		if err != nil {
+			return handleError(err)
+		}
+		rootIDBytes := make([]byte, StorageIDSize)
+		_, err := m.StorageID().ToRawBytes(rootIDBytes)
+		if err != nil {
+			return handleError(err)
+		}
+		err = db.led.SetValue(db.flowEVMRootAddress.Bytes(), []byte(FlowEVMRootSlabKey), rootIDBytes[:])
```
[the remark this answers is not in the document]
Comment (ramtinms) [reply fvm/evm/emulator/database/database.go:87]: Right, did a small refactor, added updating the ledger as part of the Database.Commit().

File: fvm/evm/emulator/database/database.go [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+
+// Len returns the number of entries currently present in the memory database.
+//
+// Note, this method is only used for testing (i.e. not public in general) and
+// does not have explicit checks for closed-ness to allow simpler testing code.
+func (db *Database) Len() int {
+	db.lock.RLock()
+	defer db.lock.RUnlock()
+
+	return db.storage.Count()
```
[the remark this answers is not in the document]
Comment (ramtinms) [reply fvm/evm/emulator/database/database.go]: nice catch, this was renamed but forgot to update it here.

File: fvm/evm/emulator/database/database.go (hunk 1 above)
Comment (ramtinms) [reply fvm/evm/emulator/database/database.go:351]: I think we don't use this functionality in our code yet, so I added as a TODO for future improvements

File: fvm/evm/emulator/database/storable.go (hunk 2 above)
Comment (ramtinms) [reply fvm/evm/emulator/database/storable.go]: updated

File: fvm/evm/emulator/database/database.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (fxamacker) [reply fvm/evm/emulator/database/database.go:87]: @ramtinms sorry, just a quick question about this.

> Right, did a small refactor at [164da3d](https://github.com/onflow/flow-go/pull/4925/commits/164da3d4910c5711ee7e6d47e9b93f9edfac7435), added updating the ledger as part of the Database.Commit().

Do we still want to keep the following line in `database.retrieveOrCreateMapRoot()`?
```Go
db.led.SetValue(db.flowEVMRootAddress.Bytes(), []byte(FlowEVMRootSlabKey), rootIDBytes[:])
```

It seems 164da3d4910c5711ee7e6d47e9b93f9edfac7435 moved the following line for `FlowEVMRootHashKey` to `database.Commit()`, not `FlowEVMRootSlabKey`.
```Go
db.led.SetValue(db.flowEVMRootAddress[:], []byte(FlowEVMRootHashKey), root[:])
```

**Comment by fxamacker** [comment]
Hey Ramtin, I took another look and left [a question (above)](https://github.com/onflow/flow-go/pull/4925#discussion_r1380834601) about refactor at [164da3d](https://github.com/onflow/flow-go/pull/4925/commits/164da3d4910c5711ee7e6d47e9b93f9edfac7435).

File: fvm/evm/emulator/database/database.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ramtinms) [reply fvm/evm/emulator/database/database.go]: Right, I missed that part and solved the other problem :D does something like this solve the issue? [0cdbeb8](https://github.com/onflow/flow-go/pull/4925/commits/0cdbeb86fd0156155088999489a57f29b21a95ab)

**Review by fxamacker — APPROVED** [review]
> :D does something like this solve the issue? [0cdbeb8](https://github.com/onflow/flow-go/pull/4925/commits/0cdbeb86fd0156155088999489a57f29b21a95ab)

Yes, thanks! LGTM! :+1:

**Review by janezpodhostnik — APPROVED** [review]

**Pull request merged**